CoreScriptsRoot/Modules/VR/Panel3D.lua

```diff
@@ -0,0 +1,760 @@
+		ray = ray.Unit
+
+		local Vd = planeNormal:Dot(ray.Direction)
+		if Vd == 0 then -- parallel, no intersection
+			return nil
+		end
+
+		local V0 = planeNormal:Dot(pointOnPlane - ray.Origin)
+		local t = V0 / Vd
+		if t < 0 then --plane is behind ray origin, and thus there is no intersection
+			return nil
+		end
+		
+		return ray.Origin + ray.Direction * t
+	end
+end
+--End of Util
+
+
+--Panel3D State variables
+local coreScriptMode = true
+local renderStepName = "Panel3DRenderStep-" .. game:GetService("HttpService"):GenerateGUID()
+local defaultPixelsPerStud = 64
+local pointUpCF = CFrame.Angles(math.rad(-90), math.rad(180), 0)
+local zeroVector = Vector3.new(0, 0, 0)
+local zeroVector2 = Vector2.new(0, 0)
+local turnAroundCF = CFrame.Angles(0, math.rad(180), 0)
+local fullyOpaqueAtPixelsFromEdge = 10
+local fullyTransparentAtPixelsFromEdge = 80
+
+local cursorHidden = false
+local cursorHideTime = 2.5
+
+local currentModal = nil
+local lastModal = nil
+local currentMaxDist = math.huge
+local currentClosest = nil
+local currentHeadScale = 1
+local panels = {}
+local headHeightFromFloor = 5
+local floorRotation = CFrame.new()
+local cursor = Util.Create "ImageLabel" {
+	Image = "rbxasset://textures/Cursors/Gamepad/Pointer.png",
+	Size = UDim2.new(0, 8, 0, 8),
+	BackgroundTransparency = 1,
+	ZIndex = 10
+}
+--End of Panel3D State variables
+
+
+--Panel3D Declaration and enumerations
+local Panel3D = {}
+Panel3D.Type = {
+	None = 0,
+	Floor = 1,
+	Fixed = 2,
+	HorizontalFollow = 3,
+	FixedToHead = 4
+}
+
+function Panel3D.GetHeadLookXZ(withTranslation)
+	local userHeadCF = UserInputService:GetUserCFrame(Enum.UserCFrame.Head)
+	local headLook = userHeadCF.lookVector
+	local headYaw = math.atan2(-headLook.Z, headLook.X) + math.rad(90)
+	local cf = CFrame.Angles(0, headYaw, 0)
+
+	if withTranslation then
+		cf = cf + userHeadCF.p
+	end
+	return cf
+end
+
+function Panel3D.FindContainerOf(element)
+	for i, v in pairs(panels) do
+		if v.gui and v.gui:IsAncestorOf(element) then
+			return v
+		end
+	end
+	return nil
+end
+
+function Panel3D.SetModalPanel(panel)
+	if currentModal == panel then
+		return
+	end
+	if currentModal then
+		currentModal:OnModalChanged(false)
+	end
+	if panel then
+		panel:OnModalChanged(true)
+	end
+	currentModal = panel
+end
+--End of Panel3D Declaration and enumerations
+
+
+--Cursor autohiding methods
+local cursorHidden = false
+local hasTool = false
+local lastMouseMove = tick()
+
+local function OnCharacterAdded(character)
+	hasTool = false
+	for i, v in ipairs(character:GetChildren()) do
+		if v:IsA("Tool") then
+			hasTool = true
+		end
+	end
+	character.ChildAdded:connect(function(child)
+		if child:IsA("Tool") then
+			hasTool = true
+			lastMouseMove = tick() --kick the mouse when a tool is equipped
+		end
+	end)
+	character.ChildRemoved:connect(function(child)
+		if child:IsA("Tool") then
+			hasTool = false
+		end
+	end)
+end
+spawn(function()
+	while not game.Players.LocalPlayer do wait() end
```

Comment:
We use game.Players a lot in this code snippet, perhaps we can save the service to a variable, as well as the LocalPlayer